asset/asset.go

```diff
@@ -801,20 +798,105 @@ var _ GenesisSigner = (*RawKeyGenesisSigner)(nil)
 
 // DeriveGroupKey derives an asset's group key based on an internal public
 // key descriptor, the original group asset genesis, and the asset's genesis.
-func DeriveGroupKey(genSigner GenesisSigner, rawKey keychain.KeyDescriptor,
-	initialGen Genesis, currentGen *Genesis) (*GroupKey, error) {
+func DeriveGroupKey(genSigner GenesisSigner, groupKey *GroupKey,
+	scriptLeaf *psbt.TaprootTapLeafScript, initialGen Genesis,
+	newAsset *Asset) (*GroupKey, error) {
+
+	// First, perform the final checks on the asset being authorized for
+	// group membership.
+	if newAsset == nil || newAsset.IsUnknownVersion() {
+		return nil, fmt.Errorf("genesis asset is missing or unknown " +
+			"version")
+	}
+
+	if newAsset.Genesis != initialGen && initialGen.Type != newAsset.Type {
+		return nil, fmt.Errorf("asset group type mismatch")
+	}
+
+	if scriptLeaf != nil && groupKey.TapscriptRoot == [32]byte{} {
+		return nil, fmt.Errorf("cannot script spend group key with " +
+			"empty tapscript root")
+	}
+
+	// Populate the sign descriptor with static and default fields.
+	genesisTweak := initialGen.ID()
+	signDesc := &lndclient.SignDescriptor{
+		KeyDesc:       groupKey.RawKey,
+		SingleTweak:   genesisTweak[:],
+		DoubleTweak:   nil,
+		TapTweak:      []byte{},
+		WitnessScript: []byte{},
+		HashType:      txscript.SigHashDefault,
+		InputIndex:    0,
+	}
+	if groupKey.TapscriptRoot != [32]byte{} {
+		signDesc.TapTweak = groupKey.TapscriptRoot[:]
+	}
+
+	tweakedGroupKey, err := GroupPubKey(
+		signDesc.KeyDesc.PubKey, signDesc.SingleTweak,
+		signDesc.TapTweak,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("cannot tweak group key: %w", err)
+	}
+
+	// Reimplement tapscript.InputAssetPrevOut to avoid a dependency cycle.
```

Comment:
Leaving comment here so we make sure it's addressed. Alternatively, we can just move this function out of the `asset` package and into `tapscript`?